src/Framework/ReuseableStringBuilder.cs

```diff
@@ -0,0 +1,355 @@
+#if DEBUG && ASSERT_BALANCE
+                int balance = Interlocked.Increment(ref s_getVsReleaseBalance);
+                Debug.Assert(balance == 1, "Unbalanced Get vs Release. Either forgotten Release or used from multiple threads concurrently.");
+#endif
+
+                StringBuilder? returned = Interlocked.Exchange(ref s_sharedBuilder, null);
+
+                if (returned == null)
+                {
+                    // Currently loaned out so return a new one with capacity in given bracket.
+                    // If user wants bigger capacity than maximum capacity, respect it.
+                    returned = new StringBuilder(SelectBracketedCapacity(capacity));
+#if DEBUG
+                    MSBuildEventSource.Log.ReusableStringBuilderFactoryStart(hash: returned.GetHashCode(), newCapacity:capacity, oldCapacity:0, type:"miss");
+#endif
+                }
+                else if (returned.Capacity < capacity)
+                {
+                    // It's essential we guarantee the capacity because this
+                    // may be used as a buffer to a PInvoke call.
+                    int newCapacity = SelectBracketedCapacity(capacity);
+#if DEBUG
+                    MSBuildEventSource.Log.ReusableStringBuilderFactoryStart(hash: returned.GetHashCode(), newCapacity: newCapacity, oldCapacity: returned.Capacity, type: "miss-need-bigger");
+#endif
+                    // Let the current StringBuilder be collected and create new with bracketed capacity. This way it allocates only char[newCapacity]
+                    //   otherwise it would allocate char[new_capacity_of_last_chunk] (in set_Capacity) and char[newCapacity] (in Clear).
+                    returned = new StringBuilder(newCapacity);
+                }
+                else
+                {
+#if DEBUG
+                    MSBuildEventSource.Log.ReusableStringBuilderFactoryStart(hash: returned.GetHashCode(), newCapacity: capacity, oldCapacity: returned.Capacity, type: "hit");
+#endif
+                }
+
+                return returned;
+            }
+
+            /// <summary>
+            /// Returns the shared builder for the next caller to use.
+            /// ** CALLERS, DO NOT USE THE BUILDER AFTER RELEASING IT HERE! **
+            /// </summary>
+            internal static void Release(ReuseableStringBuilder returning)
+            {
+#if DEBUG && ASSERT_BALANCE // Please define ASSERT_BALANCE if you need to analyze where we have cross thread competing usage of ReuseableStringBuilder
+                int balance = Interlocked.Decrement(ref s_getVsReleaseBalance);
+                Debug.Assert(balance == 0, "Unbalanced Get vs Release. Either forgotten Release or used from multiple threads concurrently.");
+#endif
+                FrameworkErrorUtilities.VerifyThrowInternalNull(returning._borrowedBuilder, nameof(returning._borrowedBuilder) + " can not be null.");
+
+                StringBuilder returningBuilder = returning._borrowedBuilder!;
```

Comment:
```suggestion
                StringBuilder returningBuilder = returning._borrowedBuilder;
```

The `!` can be avoided here by annotating `VerifyThrowInternalNull` as:

```
internal static void VerifyThrowInternalNull([NotNull] object? parameter, string parameterName)
```

This looks a bit weird, but it means:

- The argument may be null when it is passed in
- The argument will not be null when normal flow continues

---

As an aside (likely for another PR) if we're on C# 10 then it may be possible to use the `CallerArgumentExpressionAttribute` on `parameterName`:

```c#
internal static void VerifyThrowInternalNull([NotNull] object? parameter, [CallerArgumentExpressionAttribute] string parameterName = null)
```

Meaning the call could be simplified to:

```c#
FrameworkErrorUtilities.VerifyThrowInternalNull(returning._borrowedBuilder);
```

Reply:
I would rather address this in another PR. 
`FrameworkErrorUtilities.VerifyThrowInternalNull` needs some rethinking. It should be supporting nullable, but currently the lines:
https://github.com/dotnet/msbuild/blob/93e469310958da9f5a4feb473c3cb0ddddf22b75/src/Framework/ErrorUtilities.cs#L23
along with
https://github.com/dotnet/msbuild/blob/93e469310958da9f5a4feb473c3cb0ddddf22b75/src/Framework/ErrorUtilities.cs#L69-L72
breaks the rules "**The argument will not be null when normal flow continues**".
I would like to question usefulness of `MSBUILDDONOTTHROWINTERNAL` as this means "lets MSBuild fail at next line when accessing null ref object". 
@rainersigwald do people use it?

Reply:
I didn't realise it was conditional. Still, I think it would be ok to annotate it as though the return value is non-null given that's clearly the intent. NRT doesn't give hard guarantees. It's mostly about communicating intent in code and make bugs more obvious.

Reply:
I agree about doing this in another PR (and closing dotnet/msbuild#5163 while we're at it). I pushed up a branch with a half-hearted attempt at annotating ErrorUtilities that I put aside when hitting some of these same issues.